apis/r/R/write_seurat.R

```diff
@@ -207,6 +208,210 @@ write_soma.Assay <- function(
+      ingest_mode = ingest_mode,
+      platform_config = platform_config,
+      tiledbsoma_ctx = tiledbsoma_ctx
+    )
+  } else if (isTRUE(relative)) {
+    SOMACollectionOpen(uri = file_path(ms$uri, 'X'), mode = 'WRITE')
+  } else {
+    ms$X
+  }
+  withCallingHandlers(
+    .register_soma_object(X, soma_parent = ms, key = 'X', relative = relative),
+    existingKeyWarning = .maybe_muffle
+  )
+  on.exit(X$close(), add = TRUE, after = FALSE)
+
+  X$set_metadata(.layer_hint(SeuratObject::DefaultLayer(x)))
+
+  # Pull presence matrices from the v5 assay
+  cmat <- methods::slot(x, name = 'cells')
+  fmat <- methods::slot(x, name = 'features')
+
+  # Write `X` matrices
+  for (lyr in SeuratObject::Layers(x)) {
+    ldat <- SeuratObject::LayerData(x, layer = lyr)
+    if (!inherits(ldat, what = c("matrix", "Matrix"))) {
+      warning(warningCondition(
+        message = sprintf("%s", lyr),
+        class = "unknownMatrixTypeWarning",
+        call = str2lang("write_soma()")
+      ))
+      next
+    }
+    type <- .type_hint(ifelse(is.matrix(ldat), yes = 'matrix', no = class(ldat)))
+    if (all(fmat[, lyr]) && all(cmat[, lyr])) {
+      spdl::info("Adding '{}' matrix as '{}'", lyr, lyr)
+      tryCatch(
+        expr = {
+          arr <- write_soma(
+            x = ldat,
+            uri = lyr,
+            soma_parent = X,
+            sparse = TRUE,
+            transpose = TRUE,
+            ingest_mode = ingest_mode,
+            shape = shape,
+            key = lyr,
+            platform_config = platform_config,
+            tiledbsoma_ctx = tiledbsoma_ctx
+          )
+          arr$set_metadata(type)
+        },
+        error = function(err) {
+          if (lyr %in% SeuratObject::DefaultLayer(x)) {
+            stop(err)
+          }
+          err_to_warn(err)
+        }
+      )
+      next
+    }
+    ldat <- Matrix::t(as(ldat, "TsparseMatrix"))
+    idx <- which(cmat[, lyr])
+    jdx <- which(fmat[, lyr])
+    coo <- data.frame(
+      soma_dim_0 = bit64::as.integer64(idx[ldat@i + 1L] - 1L),
+      soma_dim_1 = bit64::as.integer64(jdx[ldat@j + 1L] - 1L),
+      soma_data = ldat@x
+    )
+    shape <- c(max(coo$soma_dim_0), max(coo$soma_dim_1)) + 1L
+    arr <- X$add_new_sparse_ndarray(
+      key = lyr,
+      type = arrow::infer_type(coo$soma_data),
+      shape = as.integer(shape)
+    )
+    arr$.__enclos_env__$private$.write_coo_dataframe(coo)
+    arr$set_metadata(.ragged_array_hint())
+    arr$set_metadata(type)
+  }
+
+  # Write feature-level meta data
+  var_df <- .df_index(
+    x = x[[]],
+    alt = 'features',
+    axis = 'var',
+    prefix = 'seurat'
+  )
+  var_df[[attr(x = var_df, which = 'index')]] <- rownames(x)
+  spdl::info("Adding feature-level meta data")
+  write_soma(
+    x = var_df,
+    uri = 'var',
+    soma_parent = ms,
+    key = 'var',
+    ingest_mode = ingest_mode,
+    platform_config = platform_config,
+    tiledbsoma_ctx = tiledbsoma_ctx
+  )
+
+  # Return
```

Comment:
I'm generally in favor of more comments than fewer -- don't get me wrong -- but this particular one doesn't seem helpful

Reply:
This is following the same pattern as used in the v3 Assay ingestor https://github.com/single-cell-data/TileDB-SOMA/blob/main/apis/r/R/write_seurat.R#L191; happy to remove if requested